pepper-apis/src/main/java/org/broadinstitute/ddp/service/FileUploadService.java

```diff
@@ -133,51 +137,70 @@ URL generateSignedUrl(String blobName, String mimeType, HttpMethod method) {
                 Storage.SignUrlOption.withExtHeaders(headers));
     }
 
+    // Convenience helper to lock file upload before verifying.
+    public Optional<VerifyResult> verifyUpload(Handle handle, long participantUserId, long uploadId) {
+        return handle.attach(FileUploadDao.class)
+                .findAndLockById(uploadId)
+                .map(upload -> verifyUpload(handle, participantUserId, upload));
+    }
+
     /**
      * Determine if file meets criteria, and update its status appropriately. E.g. file must be uploaded, matches
-     * reported file size, and belong to the participant. Files already marked uploaded will not be checked again.
+     * reported file size, and belong to the participant. Files already verified will not be checked again.
+     *
+     * <p>File upload should be "locked" to ensure its status doesn't change midway or file gets moved to a different
+     * bucket. Since the file upload is locked, caller should commit transaction as soon as possible to unblock other
+     * transactions waiting on the file upload row.
      *
      * @param handle            the database handle
      * @param participantUserId the participant who this upload will be assigned to
      * @param upload            the file upload
      * @return check result
      */
-    public CheckResult checkAndSetUploadStatus(Handle handle, long participantUserId, FileUpload upload) {
+    public VerifyResult verifyUpload(Handle handle, long participantUserId, FileUpload upload) {
         if (participantUserId != upload.getParticipantUserId()) {
-            return CheckResult.OWNER_MISMATCH;
+            return VerifyResult.OWNER_MISMATCH;
         }
 
-        if (upload.getStatus() == FileUploadStatus.AUTHORIZED) {
-            // File upload haven't been checked yet, do it now.
-            Blob blob = fetchBlob(upload.getBlobName());
+        if (!upload.isVerified()) {
```

Comment:
While working through the "file scan result handler", I was struggling with how to deal with the different statuses and this "verification" of the upload. I think the key insight here is that there's ultimately two different concepts here: 1) whether file is "verified", and 2) file status can be separate from "verified". File needs to be "uploaded" to be marked "verified" of course, but outside of that we can update file status separately. So to make things more manageable, verification is only responsible for checking `is_verified`, while "file scan result handler" will be responsible for updating file status (e.g. uploaded timestamp and scan result).

Reply:
Make sense to me that scan result is an independent status from user upload having been verified.